packages/visual-editor/src/utils/migrate.ts

```diff
@@ -0,0 +1,103 @@
+import {
+  Config,
+  Data,
+  DefaultComponentProps,
+  DefaultRootProps,
+  migrate as migratePuck,
+  transformProps,
+} from "@measured/puck";
+import { migrationRegistry as commonMigrationRegistry } from "../components/migrations/migrationRegistry.ts";
+
+export type MigrationAction =
+  | {
+      action: "removed";
+    }
+  | {
+      action: "renamed";
+      newName: string;
+    }
+  | {
+      action: "updated";
+      propTransformation: (
+        oldProps: Record<string, any>
+      ) => Record<string, any>;
+    };
+export type Migration = Record<string, MigrationAction>;
+export type MigrationRegistry = Migration[];
+
+interface RootProps extends DefaultRootProps {
+  version?: number;
+}
+
+export const migrate = (
+  data: Data<DefaultComponentProps, RootProps>,
+  migrationRegistry: MigrationRegistry = commonMigrationRegistry,
+  // config will be used with Puck 0.19
+  // eslint-disable-next-line  @typescript-eslint/no-unused-vars
+  config?: Config
+): Data => {
+  const version = data.root?.props?.version ?? 0;
+
+  // Apply puck migrations
+  data = migratePuck(data);
+
+  const migrationsToApply = migrationRegistry.slice(version);
+  if (migrationsToApply.length === 0) {
+    return data;
+  }
+
+  migrationsToApply.forEach((migration) => {
+    Object.entries(migration).forEach(([componentName, migrationAction]) => {
```

Comment:
Do we want to skip migrations that have already happened before? If there are 3 and we already ran through 2 in the past, we can skip those now right?